File: be/src/models/order.model.js
const db = require("../config/firebase");

const ORDERS_COLLECTION = "orders";

const OrderModel = {
  async create({ walletAddress, machineId, slot, productName, priceETH, txHash, commandId }) {
    const order = {
      walletAddress: walletAddress.toLowerCase(),
      machineId,
      slot,
      productName,
      priceETH,
      txHash,
      commandId: commandId || null,
      status: "confirmed", // confirmed, dispensed, failed
      createdAt: new Date().toISOString(),
      dispensedAt: null,
    };
    const ref = await db.collection(ORDERS_COLLECTION).add(order);
    return { id: ref.id, ...order };
  },

  async getByWallet(walletAddress, limit = 20) {
    const snapshot = await db
      .collection(ORDERS_COLLECTION)
      .where("walletAddress", "==", walletAddress.toLowerCase())
      .orderBy("createdAt", "desc")
      .limit(limit)
      .get();

    return snapshot.docs.map((doc) => ({ id: doc.id, ...doc.data() }));
  },

  async getByMachine(machineId, limit = 50) {
    const snapshot = await db
      .collection(ORDERS_COLLECTION)
      .where("machineId", "==", machineId)
      .orderBy("createdAt", "desc")
      .limit(limit)
      .get();

    return snapshot.docs.map((doc) => ({ id: doc.id, ...doc.data() }));
  },

  async updateStatus(orderId, status) {
    const updateData = {
      status,
      updatedAt: new Date().toISOString(),
    };
    if (status === "dispensed") {
      updateData.dispensedAt = new Date().toISOString();
    }
    await db.collection(ORDERS_COLLECTION).doc(orderId).update(updateData);
  },

  async setCommandId(orderId, commandId) {
    await db.collection(ORDERS_COLLECTION).doc(orderId).update({ commandId });
  },

  async updateDispensed(orderId) {
    await db.collection(ORDERS_COLLECTION).doc(orderId).update({
      status: "dispensed",
      dispensedAt: new Date().toISOString(),
    });
  },
};

module.exports = OrderModel;

File: mobile/lib/models/order_model.dart
class Order {
  final String id;
  final String walletAddress;
  final String machineId;
  final String slot;
  final String productName;
  final double priceETH;
  final String txHash;
  final String status; // confirmed, dispensed, failed
  final String createdAt;
  final String? dispensedAt;

  const Order({
    required this.id,
    required this.walletAddress,
    required this.machineId,
    required this.slot,
    required this.productName,
    required this.priceETH,
    required this.txHash,
    required this.status,
    required this.createdAt,
    this.dispensedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id']?.toString() ?? '',
      walletAddress: json['walletAddress']?.toString() ?? '',
      machineId: json['machineId']?.toString() ?? '',
      slot: json['slot']?.toString() ?? '',
      productName: json['productName']?.toString() ?? '',
      priceETH: (json['priceETH'] as num?)?.toDouble() ?? 0,
      txHash: json['txHash']?.toString() ?? '',
      status: json['status']?.toString() ?? 'confirmed',
      createdAt: json['createdAt']?.toString() ?? '',
      dispensedAt: json['dispensedAt']?.toString(),
    );
  }
}
